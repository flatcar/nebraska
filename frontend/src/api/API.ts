import { createContext } from 'react';
import _ from 'underscore';

import { CONFIG_STORAGE_KEY, NebraskaConfig } from '../stores/redux/features/config';
import { getToken } from '../utils/auth';
import {
  Activity,
  Application,
  Channel,
  FlatcarAction,
  Group,
  Instance,
  Instances,
  InstanceStatusHistory,
  Package,
  VersionBreakdownEntry,
} from './apiDataTypes';

type WithCount<T> = T & {
  count: number;
  totalCount: number;
};

const BASE_URL = '/api';
type REQUEST_DATA_TYPE = BodyInit | null | undefined;

function isNotNullUndefinedOrEmptyString(val: any) {
  return val !== undefined && val !== null && val !== '';
}

export default class API {
  // Applications

  static getApplications(): Promise<WithCount<{ applications: Application[] }>> {
    return API.getJSON(BASE_URL + '/apps');
  }

  static getApplication(applicationID: string): Promise<Application> {
    return API.getJSON(BASE_URL + '/apps/' + applicationID);
  }

  static deleteApplication(applicationID: string) {
    const url = BASE_URL + '/apps/' + applicationID;

    return API.doRequest('DELETE', url, '');
  }

  static createApplication(
    applicationData: Pick<Application, 'name' | 'description' | 'product_id'>,
    clonedFromAppID: string
  ): Promise<Application> {
    let url = BASE_URL + '/apps';
    if (clonedFromAppID) {
      url += '?clone_from=' + clonedFromAppID;
    }

    return API.doRequest('POST', url, JSON.stringify(applicationData));
  }

  static updateApplication(applicationData: Application): Promise<Application> {
    const url = BASE_URL + '/apps/' + applicationData.id;

    return API.doRequest('PUT', url, JSON.stringify(applicationData));
  }

  // Groups

  static getGroup(applicationID: string, groupID: string): Promise<Group> {
    return API.getJSON(BASE_URL + '/apps/' + applicationID + '/groups/' + groupID);
  }

  static deleteGroup(applicationID: string, groupID: string) {
    const url = BASE_URL + '/apps/' + applicationID + '/groups/' + groupID;

    return API.doRequest('DELETE', url, '');
  }

  static createGroup(groupData: Group): Promise<Group> {
    const url = BASE_URL + '/apps/' + groupData.application_id + '/groups';

    return API.doRequest('POST', url, JSON.stringify(groupData));
  }

  static updateGroup(groupData: Group): Promise<Group> {
    const url = BASE_URL + '/apps/' + groupData.application_id + '/groups/' + groupData.id;

    return API.doRequest('PUT', url, JSON.stringify(groupData));
  }

  static getGroupVersionCountTimeline(applicationID: string, groupID: string, duration: string) {
    return API.getJSON(
      `${BASE_URL}/apps/${applicationID}/groups/${groupID}/version_timeline?duration=${duration}`
    );
  }

  static getGroupStatusCountTimeline(applicationID: string, groupID: string, duration: string) {
    return API.getJSON(
      `${BASE_URL}/apps/${applicationID}/groups/${groupID}/status_timeline?duration=${duration}`
    );
  }

  static getGroupInstancesStats(applicationID: string, groupID: string, duration: string) {
    return API.getJSON(
      `${BASE_URL}/apps/${applicationID}/groups/${groupID}/instances_stats?duration=${duration}`
    );
  }

  static getGroupVersionBreakdown(
    applicationID: string,
    groupID: string
  ): Promise<VersionBreakdownEntry[]> {
    return API.getJSON(
      BASE_URL + '/apps/' + applicationID + '/groups/' + groupID + '/version_breakdown'
    );
  }

  // Channels

  static deleteChannel(applicationID: string, channelID: string) {
    const url = BASE_URL + '/apps/' + applicationID + '/channels/' + channelID;

    return API.doRequest('DELETE', url, '');
  }

  static createChannel(channelData: Channel): Promise<Channel> {
    const url = BASE_URL + '/apps/' + channelData.application_id + '/channels';

    return API.doRequest('POST', url, JSON.stringify(channelData));
  }

  static updateChannel(channelData: Channel): Promise<Channel> {
    const keysToRemove = ['id', 'created_ts', 'package'];
    const processedChannel = API.removeKeysFromObject(channelData, keysToRemove);
    const url = BASE_URL + '/apps/' + channelData.application_id + '/channels/' + channelData.id;

    return API.doRequest('PUT', url, JSON.stringify(processedChannel));
  }

  // Packages
  static getPackages(
    applicationID: string,
    searchTerm?: string,
    queryOptions?: {
      [key: string]: any;
    }
  ): Promise<WithCount<{ packages: Package[]; totalCount: number }>> {
    const params = new URLSearchParams();

    if (searchTerm) {
      params.append('searchVersion', searchTerm);
    }

    if (queryOptions) {
      Object.keys(queryOptions).forEach(key => {
        if (isNotNullUndefinedOrEmptyString(queryOptions[key])) {
          params.append(key, queryOptions[key]);
        }
      });
    }

    const queryStr = params.toString();

    const url = `${BASE_URL}/apps/${applicationID}/packages${queryStr ? '?' + queryStr : ''}`;
    return API.doRequest('GET', url);
  }

  static deletePackage(applicationID: string, packageID: string) {
    const url = BASE_URL + '/apps/' + applicationID + '/packages/' + packageID;

    return API.doRequest('DELETE', url, '');
  }

  static createPackage(packageData: Partial<Package>): Promise<Package> {
    const url = BASE_URL + '/apps/' + packageData.application_id + '/packages';

    return API.doRequest('POST', url, JSON.stringify(packageData));
  }

  static updatePackage(packageData: Partial<Package>): Promise<Package> {
    const keysToRemove = ['id', 'created_ts', 'package'];
    const processedPackage = API.removeKeysFromObject(packageData, keysToRemove);
    const url = BASE_URL + '/apps/' + packageData.application_id + '/packages/' + packageData.id;

    return API.doRequest('PUT', url, JSON.stringify(processedPackage));
  }

  // Instances

  static getInstances(
    applicationID: string,
    groupID: string,
    queryOptions: {
      [key: string]: any;
    } = {}
  ): Promise<Instances> {
    let url = BASE_URL + '/apps/' + applicationID + '/groups/' + groupID + '/instances';

    const params = new URLSearchParams();

    Object.keys(queryOptions).forEach(key => {
      if (isNotNullUndefinedOrEmptyString(queryOptions[key])) {
        params.append(key, queryOptions[key]);
      }
    });

    const queryStr = params.toString();
    url += queryStr ? '?' + queryStr : '';

    return API.getJSON(url);
  }

  static getInstancesCount(
    applicationID: string,
    groupID: string,
    duration: string
  ): Promise<number> {
    const url = `${BASE_URL}/apps/${applicationID}/groups/${groupID}/instancescount?duration=${duration}`;

    return new Promise(resolve =>
      API.getJSON(url).then((response: { count: number }) => resolve(response.count))
    );
  }

  static getInstance(
    applicationID: string,
    groupID: string,
    instanceID: string
  ): Promise<Instance> {
    const url =
      BASE_URL + '/apps/' + applicationID + '/groups/' + groupID + '/instances/' + instanceID;

    return API.getJSON(url);
  }

  static getInstanceStatusHistory(
    applicationID: string,
    groupID: string,
    instanceID: string
  ): Promise<InstanceStatusHistory[]> {
    const url =
      BASE_URL +
      '/apps/' +
      applicationID +
      '/groups/' +
      groupID +
      '/instances/' +
      instanceID +
      '/status_history';

    return API.getJSON(url);
  }

  static updateInstance(instanceID: string, alias: REQUEST_DATA_TYPE): Promise<Instance> {
    const url = BASE_URL + '/instances/' + instanceID;
    const params = JSON.stringify({ alias });
    return API.doRequest('PUT', url, params);
  }

  // Activity

  static getActivity(): Promise<WithCount<{ activities: Activity[] }>> {
    const currentDate = new Date();
    const now = currentDate.toISOString();
    currentDate.setDate(currentDate.getDate() - 7);
    const weekAgo = currentDate.toISOString();
    const query = '?start=' + weekAgo + '&end=' + now;
    const url = BASE_URL + '/activity' + query;

    return API.getJSON(url);
  }

  // Config

  static getConfig(): Promise<NebraskaConfig> {
    return API.doRequest('GET', '/config');
  }

  // Helpers

  static removeKeysFromObject(
    data: Package | Application | Group | FlatcarAction | Channel | Partial<Package>,
    valuesToRemove: string[]
  ) {
    return _.omit(data, valuesToRemove);
  }

  // Wrappers

  private static getAuthHeaders(): { headers: Record<string, string> } | never {
    const token = getToken();
    const nebraska_config = localStorage.getItem(CONFIG_STORAGE_KEY) || '{}';
    const config = JSON.parse(nebraska_config) || {};

    if (Object.keys(config).length > 0 && config.auth_mode === 'oidc') {
      if (!token) {
        // Reject immediately if OIDC is enabled but no token is available
        throw { status: 401 };
      }
      return {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      };
    }
    return { headers: {} };
  }

  static getJSON(url: string) {
    try {
      const { headers } = this.getAuthHeaders();
      return fetch(url, {
        headers,
      }).then(response => {
        if (!response.ok) {
          throw response;
        }
        return response.json();
      });
    } catch (error) {
      return Promise.reject(error);
    }
  }

  static doRequest(method: 'GET', url: string): Promise<any>;
  static doRequest(
    method: 'POST' | 'PUT' | 'PATCH',
    url: string,
    data: REQUEST_DATA_TYPE
  ): Promise<any>;
  static doRequest(method: 'DELETE', url: string, data: REQUEST_DATA_TYPE): Promise<Response>;

  static doRequest(method: string, url: string, data: REQUEST_DATA_TYPE = '') {
    try {
      const { headers: authHeaders } = this.getAuthHeaders();
      const headers: { [key: string]: string } = {
        'Content-Type': 'application/json',
        ...authHeaders,
      };
      let fetchConfigObject: {
        method: string;
        body?: REQUEST_DATA_TYPE;
        headers?: {
          [prop: string]: any;
        };
      } = {
        method: 'GET',
        headers,
      };

      if (method === 'DELETE') {
        fetchConfigObject = {
          method,
          headers,
        };
        return fetch(url, fetchConfigObject);
      } else {
        if (method !== 'GET') {
          fetchConfigObject = {
            method,
            headers,
            body: data,
          };
        }
      }

      return fetch(url, fetchConfigObject).then(response => {
        if (!response.ok) {
          throw response;
        }
        return response.json();
      });
    } catch (error) {
      return Promise.reject(error);
    }
  }
}

export const APIContext = createContext(API);
