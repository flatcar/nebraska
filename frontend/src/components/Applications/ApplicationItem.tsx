import ScheduleIcon from '@mui/icons-material/Schedule';
import { Box, Divider, Typography } from '@mui/material';
import Grid from '@mui/material/Grid';
import makeStyles from '@mui/styles/makeStyles';
import { useTranslation } from 'react-i18next';
import { Group } from '../../api/apiDataTypes';
import { applicationsStore } from '../../stores/Stores';
import { CardFeatureLabel, CardHeader, CardLabel } from '../common/Card/Card';
import ListItem from '../common/ListItem';
import MoreMenu from '../common/MoreMenu';
import ApplicationItemGroupsList from './ApplicationItemGroupsList';

const useStyles = makeStyles({
  root: {
    padding: '0px 8px',
  },
  itemSection: {
    padding: '0 1em',
  },
});

export interface ApplicationItemProps {
  onUpdate: (appID: string) => void;
  description?: string;
  groups: Group[] | null;
  numberOfInstances: number;
  id: string;
  productId: string;
  name: string;
}

export default function ApplicationItem(props: ApplicationItemProps) {
  const classes = useStyles();
  const { t } = useTranslation();
  const { description, groups, numberOfInstances, id, productId, name } = props;

  return (
    <ListItem className={classes.root}>
      <Grid container>
        <Grid item xs={12}>
          <CardHeader
            cardMainLinkLabel={name}
            cardMainLinkPath={{ pathname: `/apps/${id}` }}
            cardId={id}
            cardTrack={productId}
            cardDescription={description || t('applications|description_none_provided')}
          >
            <MoreMenu
              options={[
                {
                  label: t('frequent|edit'),
                  action: () => props.onUpdate(id),
                },
                {
                  label: t('frequent|delete'),
                  action: () => {
                    window.confirm(t('applications|confirm_delete_application'))
                      ? applicationsStore().deleteApplication(id)
                      : null;
                  },
                },
              ]}
            />
          </CardHeader>
        </Grid>
        <Grid item xs={12}>
          <Grid container className={classes.itemSection} spacing={0}>
            <Grid item xs={4}>
              <Box mt={2}>
                <CardFeatureLabel>{t('applications|instances_title')}</CardFeatureLabel>
                <CardLabel>
                  <Typography variant="h5">
                    {numberOfInstances || t('applications|none')}
                  </Typography>
                  <Box display="flex" my={1}>
                    <ScheduleIcon color="disabled" />
                    <Box pl={1} color="text.disabled">
                      <Typography variant="subtitle1">
                        {t('applications|time_last_24_hours')}
                      </Typography>
                    </Box>
                  </Box>
                </CardLabel>
              </Box>
            </Grid>
            <Box width="1%">
              <Divider orientation="vertical" variant="fullWidth" />
            </Box>
            <Grid item xs={7}>
              <Box mt={2} p={1}>
                <CardFeatureLabel>{t('frequent|groups')}</CardFeatureLabel>
                <Box display="inline-block" pl={2}>
                  <CardLabel>
                    {groups?.length === 0 ? t('applications|none') : groups?.length}
                  </CardLabel>
                </Box>
                <ApplicationItemGroupsList groups={groups} appID={id} appName={name} />
              </Box>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </ListItem>
  );
}
