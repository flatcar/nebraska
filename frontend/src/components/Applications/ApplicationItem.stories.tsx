import { Meta, StoryFn } from '@storybook/react';
import { JSX } from 'react/jsx-runtime';
import { MemoryRouter } from 'react-router-dom';

import API, { APIContext } from '../../api/API';
import ApplicationItem, { ApplicationItemProps } from './ApplicationItem';

export default {
  title: 'applications/ApplicationItem',
  argTypes: {
    onUpdate: { action: 'onUpdate' },
  },
} as Meta;

const Template: StoryFn<ApplicationItemProps> = (
  args: JSX.IntrinsicAttributes & ApplicationItemProps
) => {
  class APIMock extends API {
    static getInstancesCount(): Promise<number> {
      return new Promise(resolve => resolve(20));
    }
  }

  return (
    <APIContext.Provider value={APIMock}>
      <MemoryRouter>
        <ApplicationItem {...args} />
      </MemoryRouter>
    </APIContext.Provider>
  );
};

export const NoGroups = {
  render: Template,

  args: {
    numberOfInstances: 1,
    id: '123',
    productId: 'xxx-xxx-yyy',
    name: 'ABC',
    description: 'App Item Description',
    groups: [],
  },
};

export const Application = {
  render: Template,

  args: {
    numberOfInstances: 1,
    id: '123',
    productId: 'xxx-xxx-yyy',
    name: 'ABC',
    description: 'App Item Description',
    groups: [
      {
        id: 'xxx-xxx-1',
        name: 'First group',
        description: '',
        created_ts: '',
        rollout_in_progress: false,
        application_id: '',
        channel_id: null,
        policy_updates_enabled: false,
        policy_safe_mode: false,
        policy_office_hours: false,
        policy_timezone: null,
        policy_period_interval: '',
        policy_max_updates_per_period: 0,
        policy_update_timeout: '',

        channel: {
          id: 'DEF',
          name: 'main',
          color: '#777777',
          created_ts: '2018-10-16T21:07:56.819939+05:30',
          application_id: '123',
          package_id: 'XYZ',
          package: {
            id: 'PACK_ID',
            type: 4,
            version: '1.11.3',
            url: 'https://github.com/kinvolk',
            filename: '',
            description: '',
            size: '',
            hash: '',
            created_ts: '2019-07-18T20:10:39.163326+05:30',
            channels_blacklist: null,
            application_id: 'df1c8bbb-f525-49ee-8c94-3ca548b42059',
            flatcar_action: null,
            arch: 0,
            extra_files: [],
          },
          arch: 0,
        },
        track: '',
      },
    ],
  },
};
