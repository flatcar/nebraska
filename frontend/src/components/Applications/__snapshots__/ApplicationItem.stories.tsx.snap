// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications/ApplicationItem Application smoke-test 1`] = `
<li class="MuiListItem-root MuiListItem-padding MuiListItem-divider stable-makeStyles-divider-20 stable-makeStyles-root-18 css-1oz7m58-MuiListItem-root"
    data-testid="list-item"
>
  <div class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root">
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container stable-makeStyles-gridHeader-21 css-1xzqqmo-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-j1w48a-MuiGrid-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <h2 class="MuiTypography-root MuiTypography-h2 css-150084s-MuiTypography-root">
              <a class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover stable-makeStyles-mainLink-23 css-1emptwe-MuiTypography-root-MuiLink-root"
                 href="/apps/123"
              >
                ABC
              </a>
            </h2>
          </div>
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <div class="MuiBox-root css-1yq8yj7">
              <p class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap stable-makeStyles-idLabel-26 css-1s94icn-MuiTypography-root"
                 arial-label="group-id"
              >
                xxx-xxx-yyy
              </p>
            </div>
            <div class="MuiBox-root css-1yuhvjn">
              <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-descriptionLabel-25 css-1s23nr9-MuiTypography-root">
                App Item Description
              </span>
            </div>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <div>
            <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-1pux6rn-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                    aria-controls="simple-menu-2"
                    aria-haspopup="true"
                    aria-label="Open menu"
                    data-testid="more-menu-open-button"
            >
              <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                   focusable="false"
                   aria-hidden="true"
                   viewbox="0 0 24 24"
                   data-testid="MoreVertIcon"
              >
                <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
                </path>
              </svg>
              <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
              </span>
            </button>
          </div>
        </div>
      </div>
      <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-light stable-makeStyles-innerDivider-27 css-xetjfm-MuiDivider-root">
    </div>
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container stable-makeStyles-itemSection-19 css-11lq3yg-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root">
          <div class="MuiBox-root css-1yuhvjn">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-24 css-1s23nr9-MuiTypography-root">
              INSTANCES
            </span>
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-28 stable-makeStyles-label-32 css-1s23nr9-MuiTypography-root">
              <h5 class="MuiTypography-root MuiTypography-h5 css-xqgeko-MuiTypography-root">
                1
              </h5>
              <div class="MuiBox-root css-1jzjbfr">
                <svg class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                     focusable="false"
                     aria-hidden="true"
                     viewbox="0 0 24 24"
                     data-testid="ScheduleIcon"
                >
                  <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                  </path>
                  <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                  </path>
                </svg>
                <div class="MuiBox-root css-1hsc5oo">
                  <h6 class="MuiTypography-root MuiTypography-subtitle1 css-15qi99c-MuiTypography-root">
                    last 24 hours
                  </h6>
                </div>
              </div>
            </span>
          </div>
        </div>
        <div class="MuiBox-root css-1pelcm7">
          <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-ftc7wc-MuiDivider-root">
        </div>
        <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 css-uqwprf-MuiGrid-root">
          <div class="MuiBox-root css-10wfmn8">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-24 css-1s23nr9-MuiTypography-root">
              Groups
            </span>
            <div class="MuiBox-root css-1t04ftt">
              <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-28 stable-makeStyles-label-34 css-1s23nr9-MuiTypography-root">
                1
              </span>
            </div>
            <div class="MuiBox-root css-164r41r">
              <div class="MuiBox-root css-1ckupud">
                <div class="MuiBox-root css-126n10b">
                  <a class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover stable-makeStyles-groupLink-35 css-1emptwe-MuiTypography-root-MuiLink-root"
                     href="/apps//groups/xxx-xxx-1"
                  >
                    First group
                  </a>
                </div>
                <div class="MuiBox-root css-qiye2o">
                  <a class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover stable-makeStyles-instanceLink-36 css-1emptwe-MuiTypography-root-MuiLink-root"
                     href="/apps//groups/xxx-xxx-1/instances?period=1d"
                  >
                    <div class="MuiBox-root css-k008qs">
                      <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                           focusable="false"
                           aria-hidden="true"
                           viewbox="0 0 24 24"
                           data-testid="LayersOutlinedIcon"
                      >
                        <path d="m11.99 18.54-7.37-5.73L3 14.07l9 7 9-7-1.63-1.27zM12 16l7.36-5.73L21 9l-9-7-9 7 1.63 1.27zm0-11.47L17.74 9 12 13.47 6.26 9z">
                        </path>
                      </svg>
                      <div class="MuiBox-root css-7ieh0i">
                        20
                      </div>
                      instances
                    </div>
                  </a>
                </div>
              </div>
              <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1i06p90-MuiListItem-root">
                <div class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root">
                  <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
                    <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault stable-makeStyles-colorAvatar-38 stable-makeStyles-colorAvatar-39 css-1onlcy6-MuiAvatar-root">
                    </div>
                  </div>
                  <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
                    <div class="MuiListItemText-root MuiListItemText-multiline stable-makeStyles-root-37 css-konndc-MuiListItemText-root">
                      <div class="MuiBox-root css-70qvj9">
                        <div class="MuiBox-root css-1xsd608">
                          main
                        </div>
                      </div>
                      <div class="MuiBox-root css-15aqakd">
                        <div class="MuiBox-root css-0">
                          1.11.3 (ALL)
                        </div>
                        <div class="MuiBox-root css-n085mf">
                          <div class="MuiBox-root css-k008qs">
                            <div class="MuiBox-root css-0">
                              <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                                   focusable="false"
                                   aria-hidden="true"
                                   viewbox="0 0 24 24"
                                   data-testid="ScheduleIcon"
                                   aria-label="Release date"
                                   data-mui-internal-clone-element="true"
                              >
                                <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                                </path>
                                <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                                </path>
                              </svg>
                            </div>
                            <div class="MuiBox-root css-1iuj5ih">
                              07/18/2019
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</li>
`;

exports[`applications/ApplicationItem NoGroups smoke-test 1`] = `
<li class="MuiListItem-root MuiListItem-padding MuiListItem-divider stable-makeStyles-divider-3 stable-makeStyles-root-1 css-1oz7m58-MuiListItem-root"
    data-testid="list-item"
>
  <div class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root">
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container stable-makeStyles-gridHeader-4 css-1xzqqmo-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-j1w48a-MuiGrid-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <h2 class="MuiTypography-root MuiTypography-h2 css-150084s-MuiTypography-root">
              <a class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover stable-makeStyles-mainLink-6 css-1emptwe-MuiTypography-root-MuiLink-root"
                 href="/apps/123"
              >
                ABC
              </a>
            </h2>
          </div>
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
            <div class="MuiBox-root css-1yq8yj7">
              <p class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap stable-makeStyles-idLabel-9 css-1s94icn-MuiTypography-root"
                 arial-label="group-id"
              >
                xxx-xxx-yyy
              </p>
            </div>
            <div class="MuiBox-root css-1yuhvjn">
              <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-descriptionLabel-8 css-1s23nr9-MuiTypography-root">
                App Item Description
              </span>
            </div>
          </div>
        </div>
        <div class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root">
          <div>
            <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-1pux6rn-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                    aria-controls="simple-menu-1"
                    aria-haspopup="true"
                    aria-label="Open menu"
                    data-testid="more-menu-open-button"
            >
              <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                   focusable="false"
                   aria-hidden="true"
                   viewbox="0 0 24 24"
                   data-testid="MoreVertIcon"
              >
                <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
                </path>
              </svg>
              <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
              </span>
            </button>
          </div>
        </div>
      </div>
      <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-light stable-makeStyles-innerDivider-10 css-xetjfm-MuiDivider-root">
    </div>
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-container stable-makeStyles-itemSection-2 css-11lq3yg-MuiGrid-root">
        <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root">
          <div class="MuiBox-root css-1yuhvjn">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-7 css-1s23nr9-MuiTypography-root">
              INSTANCES
            </span>
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-11 stable-makeStyles-label-15 css-1s23nr9-MuiTypography-root">
              <h5 class="MuiTypography-root MuiTypography-h5 css-xqgeko-MuiTypography-root">
                1
              </h5>
              <div class="MuiBox-root css-1jzjbfr">
                <svg class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                     focusable="false"
                     aria-hidden="true"
                     viewbox="0 0 24 24"
                     data-testid="ScheduleIcon"
                >
                  <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                  </path>
                  <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                  </path>
                </svg>
                <div class="MuiBox-root css-1hsc5oo">
                  <h6 class="MuiTypography-root MuiTypography-subtitle1 css-15qi99c-MuiTypography-root">
                    last 24 hours
                  </h6>
                </div>
              </div>
            </span>
          </div>
        </div>
        <div class="MuiBox-root css-1pelcm7">
          <hr class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-ftc7wc-MuiDivider-root">
        </div>
        <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 css-uqwprf-MuiGrid-root">
          <div class="MuiBox-root css-10wfmn8">
            <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-featureLabel-7 css-1s23nr9-MuiTypography-root">
              Groups
            </span>
            <div class="MuiBox-root css-1t04ftt">
              <span class="MuiTypography-root MuiTypography-body1 stable-makeStyles-label-11 stable-makeStyles-label-17 css-1s23nr9-MuiTypography-root">
                None
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</li>
`;
